tests/coverage/mcdc_non_control_flow.coverage

```diff
@@ -0,0 +1,86 @@
+   LL|       |#![feature(coverage_attribute)]
+   LL|       |//@ edition: 2021
+   LL|       |//@ compile-flags: -Zcoverage-options=mcdc
+   LL|       |//@ llvm-cov-flags: --show-mcdc
+   LL|       |
+   LL|       |use core::hint::black_box;
+   LL|       |
+   LL|      3|fn assign_and(a: bool, b: bool) {
+   LL|      3|    let x = a && b;
+                               ^2
+   LL|      3|    black_box(x);
+   LL|      3|}
+   LL|       |
+   LL|      3|fn assign_or(a: bool, b: bool) {
+   LL|      3|    let x = a || b;
+                               ^1
+   LL|      3|    black_box(x);
+   LL|      3|}
+   LL|       |
+   LL|      4|fn assign_3(a: bool, b: bool, c: bool) {
+   LL|      4|    let x = a || b && c;
+                               ^2   ^1
+   LL|      4|    black_box(x);
```

Comment:
It took me some times to understand why this would not create an MCDC record for the first 2 conditions.

Actually, it's just because of operator's priority. The AST looks like this

```
  ||
 /  \
a   &&
   /  \
  b    c
```

Therefore, before the change, 
- We would instrument `||`'s LHS through `then_else_break`, which would not produce a record because there is only 1 condition (`a`)
- We would repeat the same for `&&`, its LHS being just `b`.

This emphasizes another weak point of the previous implementation :
```rust
let x = a && (b && ( c && (d && (e))));
```
This decision wasn't recorded at all by MCDC.

Reply:
Good catch. The existing MC/DC tests are enough to demonstrate that the feature exists and functions, but aren't very thorough about testing a variety of code patterns.

(Honestly, prior to reviewing #123409 I had no idea that `a && b && c` normally gets parsed as `(a && b) && c`; I would have assumed the opposite.)